# Amazon Dataset

In [31]:
pip install sqlalchemy mysql-connector-python

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [32]:
#Import necessay libraries
import numpy as np 
import pandas as pd
import mysql.connector
from sqlalchemy import create_engine

In [33]:
# MySQL database connection configuration
db_config = {
    'host': 'localhost',
    'user': 'root',
    'password': 'kenyan23',
    'database': 'ecommerce_db'
}


In [34]:
# Create SQLAlchemy engine using the db_config dictionary
engine = create_engine(
    f"mysql+mysqlconnector://{db_config['user']}:{db_config['password']}@{db_config['host']}/{db_config['database']}"
)

In [17]:
db_config = {
    'allow_local_infile': 'True'
}

In [26]:
# Establish connection to MySQL database
conn = mysql.connector.connect(**db_config)

In [18]:
# Function to execute MySQL queries

def execute_query(query, params=None):
    cursor = conn.cursor()
    if params:
        cursor.execute(query, params)  # Execute with parameters
    else:
        cursor.execute(query)  # Execute without parameters
    conn.commit()
    cursor.close()


## Data Preprocessing

### Loading the Dataset

In [19]:
#loading dataset
df = pd.read_csv('amazon.csv')

Exploring Dataset

In [ ]:
#Get information of the no. of rows and columns
df.shape

(1465, 16)

In [ ]:
#titles of the 12 columns
print (df.columns.values)

['product_id' 'product_name' 'category' 'discounted_price' 'actual_price'
 'discount_percentage' 'rating' 'rating_count' 'about_product' 'user_id'
 'user_name' 'review_id' 'review_title' 'review_content' 'img_link'
 'product_link']


In [ ]:
df.head(5)

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [ ]:
df.tail(10)

product_id                                       product_name  \
1455  B00GHL8VP2  USHA 1212 PTC with Adjustable Thermostat Fan H...   
1456  B0B9JZW1SQ  4 in 1 Handheld Electric Vegetable Cutter Set,...   
1457  B00TI8E7BI  Philips HD9306/06 1.5-Litre Electric Kettle (M...   
1458  B07J9KXQCC  Libra Room Heater for Home, Room Heaters Home ...   
1459  B0B3JSWG81  NGI Store 2 Pieces Pet Hair Removers for Your ...   
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...   
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

                                               category discounted_price  \
1455  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...        ₹3,487.77   
1456  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...             ₹498   
1457  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,695   
1458  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...             ₹949   
1459  Home&Kitchen|Kitchen&HomeAppliances|Vacuum,Cle...             ₹199   
1460  Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...             ₹379   
1461  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,280   
1462  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...           ₹2,219   
1463  Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...           ₹1,399   
1464  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,863   

     actual_price discount_percentage rating rating_count  \
1455       ₹4,990                 30%    4.1        1,127   
1456       ₹1,200                 59%    3.2          113   
1457       ₹2,695                  0%    4.4        2,518   
1458       ₹2,299                 59%    3.6          550   
1459         ₹999                 80%    3.1            2   
1460         ₹919                 59%      4        1,090   
1461       ₹3,045                 25%    4.1        4,118   
1462       ₹3,080                 28%    3.6          468   
1463       ₹1,890                 26%      4        8,031   
1464       ₹3,690                 22%    4.3        6,987   

                                          about_product  \
1455  Meant for Spot Heating|Room Size: Upto 15 sq f...   
1456  { 4 in 1 multi-function Electric Vegetable Cut...   
1457  Food grade stainless steel, Frequency: 50 Hz. ...   
1458  Libra room heater for home has cool, warm or h...   
1459  Versatile, Protects Your Clothes: MoveCatcher ...   
1460  SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...   
1461                       230 Volts, 400 watts, 1 Year   
1462  International design and styling|Two heat sett...   
1463  Fan sweep area: 230 MM ; Noise level: (40 - 45...   
1464  Brand-Borosil, Specification – 23V ~ 5Hz;1 W...   

                                                user_id  \
1455  AECFYIUCHSZXDLACTYPEUSM5DIKA,AF546S43YKYCJWRD7...   
1456  AFCTMQKPVJI6Y2JPIGDKRKIAV43A,AF6XUHN32GSFA7LFG...   
1457  AHYXOMUJUKZHBWHP43ZAB265EDGA,AG4C27NTZZ7HTG6W3...   
1458  AE5FZ5B3EEES45Q26PNUBTJ5DRYA,AGL2B432J2VDHOSUW...   
1459                       AFPKVN5KLHB4MHSYDS25Q5MIWWKQ   
1460  AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...   
1461  AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...   
1462  AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY4SR7YQZ...   
1463  AF2JQCLSCY3QJATWUNNHUSVUPNQQ,AFDMLUXC5LS5RXDJS...   
1464  AFGW5PT3R6ZAVQR4Y5MWVAKBZAYA,AG7QNJ2SCS5VS5VYY...   

                                              user_name  \
1455  Manoj Kumar,Arun Thottakathu Prasannakumar,son...   
1456  Ashish,Kavita J.,Sundar,Tessy S.,saurabh manro...   
1457  AmazonCust,DD_Zen,Amazon Customer,VINISH THOMA...   
1458  gitesh jaiswal,Divyansh,Dr Paras,Amazon Custom...   
1459                                          Kunal Sen   
1460  Prabha ds,Raghuram bk,Real Deal,Amazon Custome...   
14

In [ ]:
df.describe()

product_id                                       product_name  \
count         1465                                               1465   
unique        1351                                               1337   
top     B07JW9H4J1  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
freq             3                                                  5   

                                                 category discounted_price  \
count                                                1465             1465   
unique                                                211              550   
top     Computers&Accessories|Accessories&Peripherals|...             ₹199   
freq                                                  233               53   

       actual_price discount_percentage rating rating_count  \
count          1465                1465   1465         1463   
unique          449                  92     28         1143   
top            ₹999                 50%    4.1        9,378   
freq            120                  56    244            9   

                                            about_product  \
count                                                1465   
unique                                               1293   
top     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
freq                                                    6   

                                                  user_id  \
count                                                1465   
unique                                               1194   
top     AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   
freq                                                   10   

                                                user_name  \
count                                                1465   
unique                                               1194   
top     $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   
freq                                                   10   

                                                review_id  \
count                                                1465   
unique                                               1194   
top     R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   
freq                                                   10   

                                             review_title  \
count                                                1465   
unique                                               1194   
top     Worked on iPhone 7 and didn’t work on XR,Good ...   
freq                                                   10   

                                           review_content  \
count                                                1465   
unique                                               1212   
top     I am not big on camera usage, personally. I wa...   
freq                                                    8   

                                                 img_link  \
count                                                1465   
unique                                               1412   
top     https://m.media-amazon.com/images/I/413sCRKobN...   
freq                                                    3   

                                             product_link  
count                                                1465  
unique                                               1465  
top     https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
freq                                                    1

In [37]:
#checking for missing values 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [95]:
#rating count has less values
# Using median to fill in the existing values 
# fill in missing values with the mean
# Remove commas (if numbers are formatted with commas)
df1 = df.copy()
# Remove commas and convert 'rating_count' to numeric safely
df.loc[:, 'rating_count'] = df['rating_count'].replace(',', '', regex=True)
df.loc[:, 'rating_count'] = pd.to_numeric(df['rating_count'], errors='coerce')

In [96]:
# Identify non-numeric values in the 'rating' column
invalid_ratings = df1[~df1["rating"].astype(str).str.replace(".", "", 1).str.isdigit()]["rating"].unique()
print("Invalid ratings found:", invalid_ratings)


Invalid ratings found: []


In [97]:
# Convert 'rating' to numeric, setting errors='coerce' will turn invalid values into NaN
df1["rating"] = pd.to_numeric(df1["rating"], errors="coerce")

# Verify that the column is now numeric
print(df1["rating"].dtype)


float64


In [99]:
df1['rating_count'].fillna(df1['rating_count'].mean(), inplace=True)


# Verify cleaning
print(df1.info())  # Check remaining data


<class 'pandas.core.frame.DataFrame'>
Index: 1464 entries, 0 to 1464
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1464 non-null   object 
 1   product_name         1464 non-null   object 
 2   category             1464 non-null   object 
 3   discounted_price     1464 non-null   float64
 4   actual_price         1464 non-null   float64
 5   discount_percentage  1464 non-null   int64  
 6   rating               1464 non-null   float64
 7   rating_count         1464 non-null   float64
 8   about_product        1464 non-null   object 
 9   user_id              1464 non-null   object 
 10  user_name            1464 non-null   object 
 11  review_id            1464 non-null   object 
 12  review_title         1464 non-null   object 
 13  review_content       1464 non-null   object 
 14  img_link             1464 non-null   object 
 15  product_link         1464 non-null   object

/tmp/ipykernel_9735/1286867660.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df1['rating_count'].fillna(df1['rating_count'].mean(), inplace=True)


In [92]:
mean_rating_count = df['rating_count'].mean()
print("Mean rating count:", mean_rating_count)

Mean rating count: 18307.376880984953


In [93]:
#confirming if missing values are handled
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1464 entries, 0 to 1464
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1464 non-null   object 
 1   product_name         1464 non-null   object 
 2   category             1464 non-null   object 
 3   discounted_price     1464 non-null   float64
 4   actual_price         1464 non-null   float64
 5   discount_percentage  1464 non-null   int64  
 6   rating               1464 non-null   float64
 7   rating_count         1462 non-null   float64
 8   about_product        1464 non-null   object 
 9   user_id              1464 non-null   object 
 10  user_name            1464 non-null   object 
 11  review_id            1464 non-null   object 
 12  review_title         1464 non-null   object 
 13  review_content       1464 non-null   object 
 14  img_link             1464 non-null   object 
 15  product_link         1464 non-null   object

Missing values are handled well and all columns have 1,465 values

In [63]:
# Check if the primary key column has duplicates
duplicates = df[df.duplicated(subset=['product_id'], keep=False)]  # Change 'id' to your primary key column

# Display duplicate entries
print(duplicates)


      product_id                                       product_name  \
0     B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1     B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2     B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3     B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4     B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   
...          ...                                                ...   
1007  B0B9BXKBC7  WeCool S5 Long Selfie Stick, with Large Reinfo...   
1010  B07GVGTSLN  Wayona Usb Type C Fast Charger Cable Fast Char...   
1017  B0B9BD2YL4  KINGONE Wireless Charging Pencil (2nd Generati...   
1018  B0BMXMLSMM  Lapster 65W compatible for OnePlus Dash Warp C...   
1019  B0141EZMAI  Gizga Essentials USB WiFi Adapter for PC, 150 ...   

                                               category  discounted_price  \
0     Computers&Accessories|Accessories&Peripherals|...             39

### Data Cleaning

In [41]:
# Display data types
print(df[['discounted_price', 'actual_price']].dtypes)

# Display first few rows to check for symbols
print(df[['discounted_price', 'actual_price']].head())


discounted_price    object
actual_price        object
dtype: object
  discounted_price actual_price
0             ₹399       ₹1,099
1             ₹199         ₹349
2             ₹199       ₹1,899
3             ₹329         ₹699
4             ₹154         ₹399


In [42]:
# Remove currency symbols and commas, then convert to numeric
df['discounted_price'] = df['discounted_price'].replace('[^\d.]', '', regex=True)
df['actual_price'] = df['actual_price'].replace('[^\d.]', '', regex=True)

# Convert to numeric type
df['discounted_price'] = pd.to_numeric(df['discounted_price'], errors='coerce')
df['actual_price'] = pd.to_numeric(df['actual_price'], errors='coerce')


In [43]:
# Confirm data types
print(df[['discounted_price', 'actual_price']].dtypes)

# Check for NaN values (if any errors occurred)
print(df[['discounted_price', 'actual_price']].isnull().sum())

# Display first few rows
print(df[['discounted_price', 'actual_price']].head())


discounted_price    float64
actual_price        float64
dtype: object
discounted_price    0
actual_price        0
dtype: int64
   discounted_price  actual_price
0             399.0        1099.0
1             199.0         349.0
2             199.0        1899.0
3             329.0         699.0
4             154.0         399.0


Dicounted and actual price has been converted from object type to float64. In addition, all commas and currencies were removed from the columns. There are no missing values in the two columns

In [44]:
# Display unique values in the 'category' column
print(df['category'].unique()[:10])

['Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables'
 'Computers&Accessories|NetworkingDevices|NetworkAdapters|WirelessUSBAdapters'
 'Electronics|HomeTheater,TV&Video|Accessories|Cables|HDMICables'
 'Electronics|HomeTheater,TV&Video|Televisions|SmartTelevisions'
 'Electronics|HomeTheater,TV&Video|Accessories|RemoteControls'
 'Electronics|HomeTheater,TV&Video|Televisions|StandardTelevisions'
 'Electronics|HomeTheater,TV&Video|Accessories|TVMounts,Stands&Turntables|TVWall&CeilingMounts'
 'Electronics|HomeTheater,TV&Video|Accessories|Cables|RCACables'
 'Electronics|HomeAudio|Accessories|SpeakerAccessories|Mounts'
 'Electronics|HomeTheater,TV&Video|Accessories|Cables|OpticalCables']


In [45]:
# Extract the first category before the first "|"
df['main_category'] = df['category'].str.split('|').str[0]



In [46]:
# Standardize main category names
category_mapping = {
    "Computers&Accessories": "Computers",
    "Electronics": "Electronics",
    "HomeAudio": "Audio",
    "HomeTheater,TV&Video": "Home Theater",
}

# Apply the mapping
df['main_category'] = df['main_category'].replace(category_mapping)


In [47]:
# Display unique extracted categories
print(df['main_category'].unique())

# Count products per category
print(df['main_category'].value_counts())

['Computers' 'Electronics' 'MusicalInstruments' 'OfficeProducts'
 'Home&Kitchen' 'HomeImprovement' 'Toys&Games' 'Car&Motorbike'
 'Health&PersonalCare']
main_category
Electronics            526
Computers              453
Home&Kitchen           448
OfficeProducts          31
MusicalInstruments       2
HomeImprovement          2
Toys&Games               1
Car&Motorbike            1
Health&PersonalCare      1
Name: count, dtype: int64


I have extracted the first category and went on to clean it further. I have standardized the values and mapped it. I have then displayed it showing that there are 526 Electronics, and 453 computers etc. 

### Data Transformation

In [48]:
#creating a new column
df['discount_amount'] = df['actual_price'] - df['discounted_price']
print(df.columns)

Index(['product_id', 'product_name', 'category', 'discounted_price',
       'actual_price', 'discount_percentage', 'rating', 'rating_count',
       'about_product', 'user_id', 'user_name', 'review_id', 'review_title',
       'review_content', 'img_link', 'product_link', 'main_category',
       'discount_amount'],
      dtype='object')


In [49]:
# Remove '%' symbol and convert to float
df['discount_percentage'] = df['discount_percentage'].str.replace('%', '', regex=True)
df['discount_percentage'] = pd.to_numeric(df['discount_percentage'], errors='coerce')

In [50]:
#creating a copy of the cleaned csv 
df.head(10)

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   
5  B08Y1TFSP6  pTron Solero TB301 3A Type-C Data and Fast Cha...   
6  B08WRWPM22  boAt Micro USB 55 Tangle-free, Sturdy Micro US...   
7  B08DDRGWTJ             MI Usb Type-C Cable Smartphone (Black)   
8  B008IFXQFU  TP-Link USB WiFi Adapter for PC(TL-WN725N), N1...   
9  B082LZGK39  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   

                                            category  discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...            399.00   
1  Computers&Accessories|Accessories&Peripherals|...            199.00   
2  Computers&Accessories|Accessories&Peripherals|...            199.00   
3  Computers&Accessories|Accessories&Peripherals|...            329.00   
4  Computers&Accessories|Accessories&Peripherals|...            154.00   
5  Computers&Accessories|Accessories&Peripherals|...            149.00   
6  Computers&Accessories|Accessories&Peripherals|...            176.63   
7  Computers&Accessories|Accessories&Peripherals|...            229.00   
8  Computers&Accessories|NetworkingDevices|Networ...            499.00   
9  Computers&Accessories|Accessories&Peripherals|...            199.00   

   actual_price  discount_percentage rating  rating_count  \
0        1099.0                   64    4.2       24269.0   
1         349.0                   43    4.0       43994.0   
2        1899.0                   90    3.9        7928.0   
3         699.0                   53    4.2       94363.0   
4         399.0                   61    4.2       16905.0   
5        1000.0                   85    3.9       24871.0   
6         499.0                   65    4.1       15188.0   
7         299.0                   23    4.3       30411.0   
8         999.0                   50    4.2      179691.0   
9         299.0                   33    4.0       43994.0   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   
5  Fast Charging & Data Sync: Solero TB301 Type-C...   
6  It Ensures High Speed Transmission And Chargin...   
7  1m long Type-C USB Cable|Sturdy and Durable. W...   
8  USB WiFi Adapter —— Speedy wireless transmissi...   
9  Universal Compatibility – It is compatible wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   
5  AEQ2YMXSZWEOHK2EHTNLOS56YTZQ,AGRVINWECNY7323CW...   
6  AG7C6DAADCTRQJG2BRS3RIKDT52Q,AFU7BOMPVJ7Q3TTA4...   
7  AHW6E5LQ2BDYOIVLAJGDH45J5V5Q,AF74RSGCHPZITVFSZ...   
8  AGV3IEFANZCKECFGUM42MRH5FNOA,AEBO7NWCNXKT4AESA...   
9  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   
5  Jayesh,Rajesh k.,Soopy,amazon customer,Aman,Sh...   
6  Vivek kumar,Amazon Customer,SARTHAK,Chiranjeev...   
7  Pavan A H,Jayesh bagad,Shridhar,rajendran,karu...   
8  Azhar JuMan,Anirudh Sood,H

# MySQL Database Intergration

In [ ]:
# Create database if not exists
create_db_query = "CREATE DATABASE IF NOT EXISTS ecommerce_db;"
execute_query(create_db_query)


In [27]:
# Use the database
use_db_query = "USE ecommerce_db;"
execute_query(use_db_query)

In [73]:
execute_query("DROP TABLE IF EXISTS products;")


In [74]:
create_raw_data_table_query = """
CREATE TABLE IF NOT EXISTS products (
    product_id VARCHAR(50) PRIMARY KEY,
    product_name VARCHAR(500),
    category VARCHAR(255),
    discounted_price DECIMAL(10,2),
    actual_price DECIMAL(10,2),
    discount_percentage DECIMAL(5,2),
    rating DECIMAL(3,2),
    rating_count INT,
    about_product TEXT,
    user_id VARCHAR(256),
    user_name VARCHAR(255),
    review_id VARCHAR(256),
    review_title VARCHAR(500),
    review_content TEXT,
    img_link TEXT,
    product_link TEXT,
    main_category VARCHAR(255),
    discount_amount DECIMAL(10,2)
);
"""
execute_query(create_raw_data_table_query)

In [75]:
fetch_query = "SELECT * FROM products LIMIT 10"
# Load data into a Pandas DataFrame
fetched_df = pd.read_sql_query(fetch_query, conn)
fetched_df.head()


/tmp/ipykernel_9735/2077740863.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  fetched_df = pd.read_sql_query(fetch_query, conn)


Empty DataFrame
Columns: [product_id, product_name, category, discounted_price, actual_price, discount_percentage, rating, rating_count, about_product, user_id, user_name, review_id, review_title, review_content, img_link, product_link, main_category, discount_amount]
Index: []

In [76]:
df.head(10)

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   
5  B08Y1TFSP6  pTron Solero TB301 3A Type-C Data and Fast Cha...   
6  B08WRWPM22  boAt Micro USB 55 Tangle-free, Sturdy Micro US...   
7  B08DDRGWTJ             MI Usb Type-C Cable Smartphone (Black)   
8  B008IFXQFU  TP-Link USB WiFi Adapter for PC(TL-WN725N), N1...   
9  B082LZGK39  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   

                                            category  discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...            399.00   
1  Computers&Accessories|Accessories&Peripherals|...            199.00   
2  Computers&Accessories|Accessories&Peripherals|...            199.00   
3  Computers&Accessories|Accessories&Peripherals|...            329.00   
4  Computers&Accessories|Accessories&Peripherals|...            154.00   
5  Computers&Accessories|Accessories&Peripherals|...            149.00   
6  Computers&Accessories|Accessories&Peripherals|...            176.63   
7  Computers&Accessories|Accessories&Peripherals|...            229.00   
8  Computers&Accessories|NetworkingDevices|Networ...            499.00   
9  Computers&Accessories|Accessories&Peripherals|...            199.00   

   actual_price  discount_percentage rating  rating_count  \
0        1099.0                   64    4.2       24269.0   
1         349.0                   43    4.0       43994.0   
2        1899.0                   90    3.9        7928.0   
3         699.0                   53    4.2       94363.0   
4         399.0                   61    4.2       16905.0   
5        1000.0                   85    3.9       24871.0   
6         499.0                   65    4.1       15188.0   
7         299.0                   23    4.3       30411.0   
8         999.0                   50    4.2      179691.0   
9         299.0                   33    4.0       43994.0   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   
5  Fast Charging & Data Sync: Solero TB301 Type-C...   
6  It Ensures High Speed Transmission And Chargin...   
7  1m long Type-C USB Cable|Sturdy and Durable. W...   
8  USB WiFi Adapter —— Speedy wireless transmissi...   
9  Universal Compatibility – It is compatible wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   
5  AEQ2YMXSZWEOHK2EHTNLOS56YTZQ,AGRVINWECNY7323CW...   
6  AG7C6DAADCTRQJG2BRS3RIKDT52Q,AFU7BOMPVJ7Q3TTA4...   
7  AHW6E5LQ2BDYOIVLAJGDH45J5V5Q,AF74RSGCHPZITVFSZ...   
8  AGV3IEFANZCKECFGUM42MRH5FNOA,AEBO7NWCNXKT4AESA...   
9  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   
5  Jayesh,Rajesh k.,Soopy,amazon customer,Aman,Sh...   
6  Vivek kumar,Amazon Customer,SARTHAK,Chiranjeev...   
7  Pavan A H,Jayesh bagad,Shridhar,rajendran,karu...   
8  Azhar JuMan,Anirudh Sood,H

In [77]:
# Remove duplicate rows based on the primary key column ('id')
df_cleaned = df.drop_duplicates(subset=['product_id'], keep='first')  # Keeps the first occurrence

# Save the cleaned CSV
cleaned_csv = "/home/sharry/Documents/KCB/Teaching/Amazon/cleaned_no_duplicates.csv"
df_cleaned.to_csv(cleaned_csv, index=False)

print(f"✅ Duplicates removed! Cleaned CSV saved at: {cleaned_csv}")


✅ Duplicates removed! Cleaned CSV saved at: /home/sharry/Documents/KCB/Teaching/Amazon/cleaned_no_duplicates.csv


In [78]:
# Load the cleaned CSV file
cleaned_csv_path = "/home/sharry/Documents/KCB/Teaching/Amazon/cleaned_no_duplicates.csv"
df_cleaned = pd.read_csv(cleaned_csv_path)

# Table name in MySQL
table_name = "products"

# Save cleaned DataFrame to MySQL
df_cleaned.to_sql(name=table_name, con=engine, if_exists="append", index=False)

print("✅ Cleaned CSV successfully loaded into MySQL using SQLAlchemy!")


DatabaseError: (mysql.connector.errors.DatabaseError) 1366 (HY000): Incorrect decimal value: '|' for column 'rating' at row 1166
[SQL: INSERT INTO products (product_id, product_name, category, discounted_price, actual_price, discount_percentage, rating, rating_count, about_product, user_id, user_name, review_id, review_title, review_content, img_link, product_link, main_category, discount_amount) VALUES (%(product_id)s, %(product_name)s, %(category)s, %(discounted_price)s, %(actual_price)s, %(discount_percentage)s, %(rating)s, %(rating_count)s, %(about_product)s, %(user_id)s, %(user_name)s, %(review_id)s, %(review_title)s, %(review_content)s, %(img_link)s, %(product_link)s, %(main_category)s, %(discount_amount)s)]
[parameters: [{'product_id': 'B07JW9H4J1', 'product_name': 'Wayona Nylon Braided USB to Lightning Fast Charging and Data Sync Cable Compatible for iPhone 13, 12,11, X, 8, 7, 6, 5, iPad Air, Pro, Mini (3 FT Pack of 1, Grey)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 399.0, 'actual_price': 1099.0, 'discount_percentage': 64, 'rating': '4.2', 'rating_count': 24269.0, 'about_product': "High Compatibility : Compatible With iPhone 12, 11, X/XsMax/Xr ,iPhone 8/8 Plus,iPhone 7/7 Plus,iPhone 6s/6s Plus,iPhone 6/6 Plus,iPhone 5/5s/5c/se,i ... (662 characters truncated) ... iendly customer services, ensures the long-time enjoyment of your purchase. If you meet any question or problem, please don't hesitate to contact us.", 'user_id': 'AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBBSNLYT3ONILA,AHCTC6ULH4XB6YHDY6PCH2R772LQ,AGYHHIERNXKA6P5T7CZLXKVPT7IQ,AG4OGOFWXJZTQ2HKYIOCOY3KXF2Q,AENGU523SXMOS7JPDTW52PNNVWGQ,AEQJHCVTNINBS4FKTBGQRQTGTE5Q,AFC3FFC5PKFF5PMA52S3VCHOZ5FQ', 'user_name': 'Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jaspreet singh,Khaja moin,Anand,S.ARUMUGAM', 'review_id': 'R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1KD19VHEDV0OR,R3C02RMYQMK6FC,R39GQRVBUZBWGY,R2K9EDOE15QIRJ,R3OI7YT648TL8I', 'review_title': 'Satisfied,Charging is really fast,Value for money,Product review,Good quality,Good product,Good Product,As of now seems good', 'review_content': 'Looks durable Charging is fine tooNo complains,Charging is really fast, good product.,Till now satisfied with the quality.,This is a good product . T ... (185 characters truncated) ... rked well till date and was having no issue.Cable is also sturdy enough...Have asked for replacement and company is doing the same...,Value for money', 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T1/images/I/51UsScvHQNL._SX300_SY300_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/Wayona-Braided-WN3LG1-Syncing-Charging/dp/B07JW9H4J1/ref=sr_1_1?qid=1672909124&s=electronics&sr=1-1', 'main_category': 'Computers', 'discount_amount': 700.0}, {'product_id': 'B098NS6PVG', 'product_name': 'Ambrane Unbreakable 60W / 3A Fast Charging 1.5m Braided Type C Cable for Smartphones, Tablets, Laptops & other Type C devices, PD Technology, 480Mbps Data Sync, Quick Charge 3.0 (RCT15A, Black)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 199.0, 'actual_price': 349.0, 'discount_percentage': 43, 'rating': '4.0', 'rating_count': 43994.0, 'about_product': 'Compatible with all Type C enabled devices, be it an android smartphone (Mi, Samsung, Oppo, Vivo, Realme, OnePlus, etc), tablet, laptop (Macbook, Chr ... (307 characters truncated) ... long like a 2meters cable|Supports maximum 3A fast charging and 480 Mbps data transfer speed|6 months manufacturer warranty from the date of purchase', 'user_id': 'AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBXNGXZJT525AQ,AHONIZU3ICIEHQIGQ6R2VFRSBXOQ,AFPHD2CRPDZMWMBL7WXRSVYWS5JA,AEZ346GX3HJ4O4XNRPHCNHXQURMQ,AEPSWFPNECKO34PUC7I56ITGXR6Q,AHWVEHR5DYLVFTO2KF3IZATFQSWQ,AH4QT33M55677I7ISQOAKEQWACYQ', 'user_name': 'ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Placeholder,BharanI,sonia,Niam', 'review_id': 'RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RYGGS0M09S3KY,R17KQRUTAN5DKS,R3AAQGS6HP2QUK,R1HDNOG6TO2CCA,R3PHKXYA5AFEOU', 'review_title': 'A Good Braided Cable for Your Type C Device,Good quality product from ambrane,Super cable,As,Good quality,Good product,its good,Good quality for the price but one issue with my unit', 'review_content': "I ordered this cable to connect my phone to Android Auto of car. The cable is really strong and the connection ports are really well made. I already  ... (888 characters truncated) ... able, pretty impressive for the price.Can't blame the manufacturer.But quality issues by the distributor, they might have stored in very humid place.", 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/31zOsqQOAOL._SY445_SX342_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/Ambrane-Unbreakable-Charging-Braided-Cable/dp/B098NS6PVG/ref=sr_1_2?qid=1672909124&s=electronics&sr=1-2', 'main_category': 'Computers', 'discount_amount': 150.0}, {'product_id': 'B096MSW6CT', 'product_name': 'Sounce Fast Phone Charging Cable & Data Sync USB Cable Compatible for iPhone 13, 12,11, X, 8, 7, 6, 5, iPad Air, Pro, Mini & iOS Devices', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 199.0, 'actual_price': 1899.0, 'discount_percentage': 90, 'rating': '3.9', 'rating_count': 7928.0, 'about_product': "【 Fast Charger& Data Sync】-With built-in safety proctections and four-core copper wires promote maximum signal quality and strength and enhance charg ... (793 characters truncated) ... ou within 24 hours ! you can purchase with confidence,and every sale includes a 365-day worry-free Service to prove the importance we set on quality.", 'user_id': 'AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQLWQOWZ4N3OA,AHTPQRIMGUD4BYR5YIHBH3CCGEFQ,AEUVWXYP5LT7PZLLZENEO2NODPBQ,AHC7MPW55DOO6WNCOQVA2VHOD26A,AFDI6FRPFBTNBG7BAEB7JDJSMKDQ,AFQKCEEEKXCOHTDG4WUN3XPPHJQQ,AHKUUFNMBZIDLSSPA4FEHIO2EC7Q', 'user_name': 'Kunal,Himanshu,viswanath,sai niharka,saqib malik,Aashiq,Ramu Challa,Sanjay gupta', 'review_id': 'R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R25X4TBMPY91LX,R27OK7G99VK0TR,R207CYDCHJJTCJ,R3PCU8XMU173BT,R1IMONDOWRNU5V', 'review_title': "Good speed for earlier versions,Good Product,Working good,Good for the price,Good,Worth for money,Working nice,it's a really nice product", 'review_content': "Not quite durable and sturdy,https://m.media-amazon.com/images/W/WEBP_402378-T1/images/I/71rIggrbUCL._SY88.jpg,Working good,https://m.media-amazon.com/images/W/WEBP_402378-T1/images/I/61bKp9YO6wL._SY88.jpg,Product,Very nice product,Working well,It's a really nice product", 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T1/images/I/31IvNJZnmdL._SY445_SX342_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/Sounce-iPhone-Charging-Compatible-Devices/dp/B096MSW6CT/ref=sr_1_3?qid=1672909124&s=electronics&sr=1-3', 'main_category': 'Computers', 'discount_amount': 1700.0}, {'product_id': 'B08HDJ86NZ', 'product_name': 'boAt Deuce USB 300 2 in 1 Type-C & Micro USB Stress Resistant, Tangle-Free, Sturdy Cable with 3A Fast Charging & 480mbps Data Transmission, 10000+ Bends Lifespan and Extended 1.5m Length(Martian Red)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 329.0, 'actual_price': 699.0, 'discount_percentage': 53, 'rating': '4.2', 'rating_count': 94363.0, 'about_product': 'The boAt Deuce USB 300 2 in 1 cable is compatible with smartphones, tablets, PC peripherals, Bluetooth speakers, power banks and all other devices wi ... (401 characters truncated) ... mlessly|Deuce USB 300 cable offers a perfect 1.5 meters in length for smooth & hassle-free user experience|2 years warranty from the date of purchase', 'user_id': 'AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5SGCUQBP7YSCA,AH725ST5NW2Y4JZPKUNTIJCUK2BA,AHV3TXIFCJPMS4D5JATCEUR266MQ,AGWIGDEMFIIUAOXYY2QATNBSUGHA,AFSTSLQUV4EVEXWKBOLEFHL2H5YQ,AGAKDNBHY2FKX7I4ACRGILU7QL7A,AFNWJUWJRHCC6HN52KMG5AKZY37Q', 'user_name': 'Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh chouhan,Ravi Siddan,Himanshu Goel,Udaykumar', 'review_id': 'R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1CLH2ULIVG5U3,R2DMKIBGFKBD6R,RC89B5IAJUTR5,R3B3DDON5FH8DS,R13WAEJDI5RS36', 'review_title': "Good product,Good one,Nice,Really nice product,Very first time change,Good,Fine product but could be better,Very nice it's charging like jet", 'review_content': 'Good product,long wire,Charges good,Nice,I bought this cable for Rs.339 worthy product for this price, i tested it in various charger adapters 33w an ... (145 characters truncated) ... nty you need to go very far not practical for such a cost and mine micro to type c connector stopped working after few days only.,I like this product', 'img_link': 'https://m.media-amazon.com/images/I/41V5FtEWPkL._SX300_SY300_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/Deuce-300-Resistant-Tangle-Free-Transmission/dp/B08HDJ86NZ/ref=sr_1_4?qid=1672909124&s=electronics&sr=1-4', 'main_category': 'Computers', 'discount_amount': 370.0}, {'product_id': 'B08CF3B7N1', 'product_name': 'Portronics Konnect L 1.2M Fast Charging 3A 8 Pin USB Cable with Charge & Sync Function for iPhone, iPad (Grey)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 154.0, 'actual_price': 399.0, 'discount_percentage': 61, 'rating': '4.2', 'rating_count': 16905.0, 'about_product': '[CHARGE & SYNC FUNCTION]- This cable comes with charging & Data sync function|[HIGH QUALITY MATERIAL]- TPE + Nylon Material to make sure that the lif ... (276 characters truncated) ... rformance while using|[UNIVERSAL COMPATIBILITY]- Compatible with all devices like iPhone XS, X, XR, 8, 7, 6S, 6, 5S, iPad Pro, iPad mini and iPad Air', 'user_id': 'AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZHM4QB2KPKFUQ,AFK4NJOLFSJGWLOJIUIAROJF6YVA,AFUOTYRFUXVPEBGIXVZZ7DR3CZUA,AFDLRSXKDZ6U3U3KD46SQLFGZQRA,AH5VLM66SIK7J3IRG4NY7XVOQ55A,AE3MQNNHHLUHXURL5S7IAR7JTGNQ,AFSEOFZY67MYC7UAJU264Z5NFTLA', 'user_name': 'rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK,Bhargav,Durai Vignesh,Amazon Customer', 'review_id': 'R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R3KK4GT934ST3I,RCFHMWUSBIJO,RDO7DACXMAJ84,R3A6MEZL3LY66Z,R1ESIEKPGAYA29', 'review_title': 'As good as original,Decent,Good one for secondary use,Best quality,GOOD,Amazing product at a mind blowing price!,Nice Quality,Good product', 'review_content': "Bought this instead of original apple, does the work for 150rs, not as fast as apple charger but its a good option if you want cheap and good product ... (2153 characters truncated) ...  cable for emergency purpose that much I love this cable. Buy for this cable only emergency uses only since Good one,Good product and value for money", 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/31VzNhhqifL._SX300_SY300_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/Portronics-Konnect-POR-1080-Charging-Function/dp/B08CF3B7N1/ref=sr_1_5?qid=1672909124&s=electronics&sr=1-5', 'main_category': 'Computers', 'discount_amount': 245.0}, {'product_id': 'B08Y1TFSP6', 'product_name': 'pTron Solero TB301 3A Type-C Data and Fast Charging Cable, Made in India, 480Mbps Data Sync, Strong and Durable 1.5-Meter Nylon Braided USB Cable for Type-C Devices for Charging Adapter (Black)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 149.0, 'actual_price': 1000.0, 'discount_percentage': 85, 'rating': '3.9', 'rating_count': 24871.0, 'about_product': 'Fast Charging & Data Sync: Solero TB301 Type-C cable supports fast charge up to 5V/3A for devices and data syncing speed up to 480Mbps.|Universal Com ... (522 characters truncated) ... nnectors: pTron Soler USB-C has passed the 5KG load test, swing test, 5000+ times connect & disconnect to ensure that there are no loose connections.', 'user_id': 'AEQ2YMXSZWEOHK2EHTNLOS56YTZQ,AGRVINWECNY7323CWFXZYYIZOFTQ,AHBAT6VLOXWGYDL57KHCNCLPXAKA,AF7NDY2H6JVYTSQOZP76GCATQ34Q,AFV7ZA733ZLME4KNLZPMPCBUNPPA,AHFAAPSY2MJ5HYOU2VQDJ7AQY4NQ,AH2WGV2PEBUTICRPBEEVKF24G5LA,AEP4MK3EKOBDKTGPJTRN5RBDIODA', 'user_name': 'Jayesh,Rajesh k.,Soopy,amazon customer,Aman,Shankar,dinesh,Chitra,Ajaybabu.O.M', 'review_id': 'R7S8ANNSDPR40,R3CLZFLHVJU26P,RFF7U7MPQFUGR,R1MV1NKC23DWPI,R11D3U0V2XKDKF,R18MP1KLUE18PC,RWGJNVEH5ZQME,R1XN72FU6Q37IH', 'review_title': "It's pretty good,Average quality,very good and useful usb cable,Good USB cable. My experience was very good it is long lasting,Good,Nice product and useful,-,Sturdy but does not support 33w charging", 'review_content': "It's a good product.,Like,Very good item strong and useful USB cableValue for moneyThanks to amazon and producer,https://m.media-amazon.com/images/I/51112ZRE-1L._SY88.jpg,Good,Nice product and useful product,-,Sturdy but does not support 33w charging", 'img_link': 'https://m.media-amazon.com/images/I/31wOPjcSxlL._SX300_SY300_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/Solero-TB301-Charging-480Mbps-1-5-Meter/dp/B08Y1TFSP6/ref=sr_1_6?qid=1672909124&s=electronics&sr=1-6', 'main_category': 'Computers', 'discount_amount': 851.0}, {'product_id': 'B08WRWPM22', 'product_name': 'boAt Micro USB 55 Tangle-free, Sturdy Micro USB Cable with 3A Fast Charging & 480mbps Data Transmission (Black)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 176.63, 'actual_price': 499.0, 'discount_percentage': 65, 'rating': '4.1', 'rating_count': 15188.0, 'about_product': 'It Ensures High Speed Transmission And Charging By Offering 3A Fast Charging And Data Transmissions With Rapid Sync At 480 Mbps|The Braided Skin Make ... (553 characters truncated) ... s, Tablets, Pc Peripherals, Bluetooth Speakers, Power Banks, Game Consoles & All Micro-Usb Enabled Devices|2 years warranty from the date of purchase', 'user_id': 'AG7C6DAADCTRQJG2BRS3RIKDT52Q,AFU7BOMPVJ7Q3TTA4G67RASTGYIQ,AER5ZGIXXVYG3AWZTRZT7M2BYCEA,AHE76XQSOLGOP5ZEKTIW6KUPDWBQ,AGXTMB2XHZBEWZ2UIX7ODZ4XTU6Q,AHNM2XVU745EDPNGUOAG74PTSNRA,AH5RWQ4S72IVLZD6O75OPCFIVDXQ,AG322TYKVPLPBDXE7ABEUK5QTALQ', 'user_name': 'Vivek kumar,Amazon Customer,SARTHAK,Chiranjeevi,V V GIRI KUMAR,Rajnandini,Akshay Talla,sudhanshu chaubey', 'review_id': 'R8E73K2KWJRDS,RSD0JTIIWQQL8,R64CRSTE9SLW1,R2FRTNIIUFJE1F,RWGNX3W7UOJ7W,R32TYHHODHTF5D,RQL9ZMQUTY7P2,R280XJ5VZUBOXV', 'review_title': 'Long durable.,good,Does not charge Lenovo m8 tab,Best charging cable,good,Boat,Product was good,1.5 m का केबल मेरे लिए बहुत ही लाभदायक है ।', 'review_content': "Build quality is good and it is comes with 2 year warranty.,Good Product,Bought it for charging my mobile & tab but it doesn't work for Lenovo m8 tab ... (296 characters truncated) ... जर उसे कर सकते है (इससे जड़ा नही) । लोकल बाजार में इसका दाम 150 है ,  अमेजन पे ये केबल मुझे 67 में मिला । गर्व से कहो हम हिंदू है , जय हिंद जय भारत ,", 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T2/images/I/41jlwEZpa5L._SX300_SY300_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/boAt-Micro-USB-Tangle-Free-Transmission/dp/B08WRWPM22/ref=sr_1_8?qid=1672909124&s=electronics&sr=1-8', 'main_category': 'Computers', 'discount_amount': 322.37}, {'product_id': 'B08DDRGWTJ', 'product_name': 'MI Usb Type-C Cable Smartphone (Black)', 'category': 'Computers&Accessories|Accessories&Peripherals|Cables&Accessories|Cables|USBCables', 'discounted_price': 229.0, 'actual_price': 299.0, 'discount_percentage': 23, 'rating': '4.3', 'rating_count': 30411.0, 'about_product': '1m long Type-C USB Cable|Sturdy and Durable. With USB cable you can transfer data with speeds of upto 480 Mbps|Upto 3A output|6months warranty|Sturdy and Durable. With USB cable you can transfer data with speeds of upto 480 Mbps|6months warranty|Up To 3A Output', 'user_id': 'AHW6E5LQ2BDYOIVLAJGDH45J5V5Q,AF74RSGCHPZITVFSZN76K6GKPICA,AHDD7ZNB47QA2JLYU53HD4ML3VNQ,AHV3ELGDSOWBYUQLXSPDCSHBQRHQ,AEJU4L3ZM2GTILSJZZSNSF6VUOIA,AFVD66VQMSHPDT3A6HBBBGKRXBZA,AELKHQXVSSG6NHXLFJLLNEFRQQUQ,AGYSMAC6V6RFJJOHG2FIRPOZ6CSQ', 'user_name': 'Pavan A H,Jayesh bagad,Shridhar,rajendran,karuppu k,Akshay wani,Abbimaniu Singh,VISHAL H BADSHE', 'review_id': 'R2X090D1YHACKR,R32ZCIH9AFNJ60,R3N57EVVG0EHAF,R3QWLE8JHROKC1,R2VTSDOOUTSQ5X,R3E6FZ75Q074KH,R1SYBQLTPFCW20,RYQT96J8HPIXE', 'review_title': 'Worth for money - suitable for Android auto,Good Product,Length,Nice,Original,Very good quay Cable support fast charging.,Original MI cable for charging upto 33 watt,I am veri happy with this product as it provide turbo charging.', 'review_content': 'Worth for money - suitable for Android auto... my purpose served in car....got it for Rs.150,Everything is Okay. But packaging is not good feeling li ... (460 characters truncated) ... it.,It provided me turbo charging. I recommend for all Redmi note 10s mobile phones. As it provide turbo charging. I am veri happy with this product.', 'img_link': 'https://m.media-amazon.com/images/I/31XO-wfGGGL._SX300_SY300_QL70_FMwebp_.jpg', 'product_link': 'https://www.amazon.in/MI-MTCY001IN-USB-Type-C-Cable/dp/B08DDRGWTJ/ref=sr_1_9?qid=1672909124&s=electronics&sr=1-9', 'main_category': 'Computers', 'discount_amount': 70.0}  ... displaying 10 of 1351 total bound parameter sets ...  {'product_id': 'B00J5DYCCA', 'product_name': 'Havells Ventil Air DSP 230mm Exhaust Fan (Pista Green)', 'category': 'Home&Kitchen|Heating,Cooling&AirQuality|Fans|ExhaustFans', 'discounted_price': 1399.0, 'actual_price': 1890.0, 'discount_percentage': 26, 'rating': '4', 'rating_count': 8031.0, 'about_product': 'Fan sweep area: 230 MM ; Noise level: (40 - 45) dB; Fan body dimension (Length x Width x Depth): 290MM x 290MM x 200MM|With bird guard to protect the ... (230 characters truncated) ... Sturdy steel bird guard with specially designed metal blade|For any functional defect, please reach out to customer care team @ Email: customercare@h', 'user_id': 'AF2JQCLSCY3QJATWUNNHUSVUPNQQ,AFDMLUXC5LS5RXDJSJJRHNBURIVQ,AHMWJ32LQF2YEADFPUML4EKXUC3Q,AFXERFNWKFFWECCPZJBRP2N23XJA,AG74HALEHHRAVFNXFHBM75RWZD6Q,AEDOS7O5DUVDYA7EHU6VERPRAFMA,AFKWDCNGBBYBDQBKMFJUVJYY77XQ,AGQCHAIK5FZZOA67FTG7HET7VVJQ', 'user_name': 'Shubham Dubey,E.GURUBARAN,Mayank S.,eusuf khan,ASHISHGAJBE,hirabhai thakarda,khageswar,A.Ramu', 'review_id': 'R39Q2Y79MM9SWK,R3079BG1NIH6MB,R29A31ZELTZNJM,RQ7XAO5UTJQZT,R223OFAZGIK4X7,R27WMZV25K3TN1,R302QB4GVL3F8T,RBZRSE5J6HCF3', 'review_title': 'Fan Speed is slow,Good quality,Good product,good,Old is gold.,Good product,Nice product,Super 💕', 'review_content': 'I have installed this in my kitchen working fine is just fan speed is very slow could have been faster but not it is slow thats the reason this exhau ... (90 characters truncated) ... EBP_402378-T1/images/I/61izko3RxDL._SY88.jpg,speed could have been little bit more,,Beat performance this product.,Connecting wire is too short,Super', 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T1/images/I/51pNg1Zy4+L._SX300_SY300_.jpg', 'product_link': 'https://www.amazon.in/Havells-Ventilair-230mm-Exhaust-Grey/dp/B00J5DYCCA/ref=sr_1_505?qid=1672923617&s=kitchen&sr=1-505', 'main_category': 'Home&Kitchen', 'discount_amount': 491.0}, {'product_id': 'B01486F4G6', 'product_name': 'Borosil Jumbo 1000-Watt Grill Sandwich Maker (Black)', 'category': 'Home&Kitchen|Kitchen&HomeAppliances|SmallKitchenAppliances|SandwichMakers', 'discounted_price': 2863.0, 'actual_price': 3690.0, 'discount_percentage': 22, 'rating': '4.3', 'rating_count': 6987.0, 'about_product': 'Brand-Borosil, Specification – 23V ~ 5Hz;1 W Power for Faster Grilling|Makes 2 Sandwiches in minutes;Stainless Steel Mirror Finish Body; Easy to us ... (93 characters truncated) ... ing|Rubber feet to avoid slipping on Countertop;Indicator Display.|Material Type: Plastic|Suitable for gifting on Diwali and other special occasions.', 'user_id': 'AFGW5PT3R6ZAVQR4Y5MWVAKBZAYA,AG7QNJ2SCS5VS5VYYBNV4HDSIJ4Q,AFUDGN5MEXLKUULNTM7Y2G5P7TYA,AHXCDNSXAESERITAFELQABFVNLCA,AGRZD6CHLCUNOLMMIMIHUCG7PIFA,AFQZVGSOSOJHKFQQMCEI4725QEKQ,AEALVGXXIP46OZVXKRUXSDWZJMEA,AGEFL3AY7YXEFZA4ZJU3LP7K7OJQ', 'user_name': 'Rajib,Ajay B,Vikas Kahol,PARDEEP,Anindya Pramanik,Vikas Singh,Harshada Pimple,Saw a.', 'review_id': 'R20RBRZ0WEUJT9,ROKIFK9R2ISSE,R30EEG2FNJSN5I,R2ZC03S4QXOW4Y,R186H8YW34BQD5,R10NC3D321N59G,REKF75G4SOAOX,R2G0ZT4JQX322I', 'review_title': 'Works perfect,Ok good product,Nice Product. Recommend it. But cleaning its exterior is cumbersome.,Excellent product✌,A good product for household use,मुझे बिल्कुल भी मजा नहीं आया और वापस कर दिया।,Best product,Good', 'review_content': 'It does it job perfectly..only issue is temp control is not perfect . U need to keep checking the browness of sandwich untill it gets upto ur choise, ... (1160 characters truncated) ...  पैकेजिंग, मैनुअल और सैंडविच मेकर की फोटोग्राफ आप सभी से शेयर की है।  आप स्वयं देख के अनुमान लगा सकते हैं।,Recommend work as expected,Its easy tp use', 'img_link': 'https://m.media-amazon.com/images/W/WEBP_402378-T1/images/I/51J2Wk-+c+L._SY300_SX300_.jpg', 'product_link': 'https://www.amazon.in/Borosil-Jumbo-1000-Watt-Grill-Sandwich/dp/B01486F4G6/ref=sr_1_506?qid=1672923617&s=kitchen&sr=1-506', 'main_category': 'Home&Kitchen', 'discount_amount': 827.0}]]
(Background on this error at: https://sqlalche.me/e/20/4xp6)